### ISSUE TITLE
signing a POST request with body as a string

### ISSUE THREAD

**Issue by marekyggdrasil** [garyburd/go-oauth#20]
Thanks for great work with your library! It helped a lot with my projects, however there is one thing I could use some help with.

In the [source](https://github.com/garyburd/go-oauth/blob/255d509a6f270332ffcaef6723047891319a2ffb/oauth/oauth.go#L57-L59) it is mentioned that methods for `post`, `get`, `put` and `delete` are easy to use but not very flexible.

I have been trying to use the low level methods to make my own custom `post` method that could send a string inside the `body` with a correct signature.

I failed due to following reasons. I would have to rewrite [SetAuthorizationHeader](https://github.com/garyburd/go-oauth/blob/255d509a6f270332ffcaef6723047891319a2ffb/oauth/oauth.go#L478), then [authorizationHeader](https://github.com/garyburd/go-oauth/blob/255d509a6f270332ffcaef6723047891319a2ffb/oauth/oauth.go#L439), then [oauthParams](https://github.com/garyburd/go-oauth/blob/255d509a6f270332ffcaef6723047891319a2ffb/oauth/oauth.go#L401). This would basically require to rewrite the most low-level part of the module (correct me if I am wrong).

There is some [information provided](https://github.com/garyburd/go-oauth/blob/255d509a6f270332ffcaef6723047891319a2ffb/oauth/oauth.go#L358-L359) about [RFC standard](https://tools.ietf.org/html/rfc5849#section-3.5.2) however if I understand it correctly it is about how to pass signature inside the body, not how to generate signature with additional information inside the body.

I would really appreciate some advises on that, also if there is no way to use this library for that I will be happy to contribute and help adding this functionality.

**Comment by garyburd** [comment]
What is the problem you are trying to solve?  Describe the data you have and what you expect to put into the request header and body.

**Comment by marekyggdrasil** [comment]
Thanks for quick feedback!

Service is Woocommerce API (based on Wordpress API) - [documentation](http://woothemes.github.io/woocommerce-rest-api-docs/#authentication-over-http)

`POST http://exampleurl.com/wc-api/v3/orders`

request headers:

```
{
      'User-Agent': 'anything here',
      'Content-Type': 'application/json',
      'Accept': 'application/json'
}
```

request body (escaped json object as a string):

```
{\"payment_method\":\"bacs\",\"payment_method_title\":\"Direct Bank Transfer\",\"set_paid\":true,\"billing\":{\"first_name\":\"John\",\"last_name\":\"Doe\",\"address_1\":\"969 Market\",\"address_2\":\"\",\"city\":\"San Francisco\",\"state\":\"CA\",\"postcode\":\"94103\",\"country\":\"US\",\"email\":\"john.doe@example.com\",\"phone\":\"(555) 555-5555\"},\"shipping\":{\"first_name\":\"John\",\"last_name\":\"Doe\",\"address_1\":\"969 Market\",\"address_2\":\"\",\"city\":\"San Francisco\",\"state\":\"CA\",\"postcode\":\"94103\",\"country\":\"US\"},\"line_items\":[{\"product_id\":12,\"quantity\":2},{\"product_id\":16,\"variation_id\":23,\"quantity\":1}],\"shipping_lines\":[{\"method_id\":\"flat_rate\",\"method_title\":\"Flat Rate\",\"total\":10}]}
```

and I have:

consumerKey: `ck_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`
consumerSecret: `cs_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`

Basically I want to be able to make requests just like [this](https://www.npmjs.com/package/woocommerce) ([source](https://github.com/woothemes/wc-api-node/blob/master/index.js#L147)) library in Node which depends on [ddo/oauth-1.0a](https://github.com/ddo/oauth-1.0a). I managed to make it sign my `GET` request, just no clue how to proceed with `POST`.
When I will make it work I will be happy to pull request an example for Woocommerce as my contribution.

**Comment by garyburd** [comment]
The oauth spec does not describe how to sign a body that is not a form. Some services handle this by ignoring the body when computing the signature.  Try this:

	req, err := http.NewRequest("POST", "http://exampleurl.com/wc-api/v3/orders", strings.NewReader(jsonText))
	if err != nil {
                // handle error
	}
	req.Header.Set("Content-Type", "application/json")
        req.Header.Set("Accept", "application/json")
        req.Header.Set("User-Agent", "anything here")
	if err := c.SetAuthorizationHeader(req.Header, credentials, "POST", req.URL, nil); err != nil {
                // handle error
	}
        resp, err := http.DefaultClient.Do(req)

**Comment by marekyggdrasil** [comment]
It worked!

I made this function based on your input. The last argument is the body string. Just one comment, it sets the authorization headers in the query string because in case of Wordpress / Woocommerce there is no reliable way to pass the headers, all libraries set the query string if possible.

```go
func (c *Client) CustomPost(client *http.Client, credentials *Credentials, urlStr string, form url.Values, data string) (*http.Response, error) {
	req, err := http.NewRequest("POST", urlStr, strings.NewReader(data))
	if err != nil {
		return nil, err
	}
	if req.URL.RawQuery != "" {
		return nil, errors.New("oauth: url must not contain a query string")
	}
	for k, v := range c.Header {
		req.Header[k] = v
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "anything here")
	if err := c.SetAuthorizationHeader(req.Header, credentials, "POST", req.URL, form); err != nil {
		return nil, err
	}
	req.URL.RawQuery = form.Encode()
	log.Println(req.URL.RawQuery)
	log.Println(form.Encode())
	if client == nil {
		client = http.DefaultClient
	}
	return client.Do(req)
}
```

Thanks a lot! Closing!
